Add function to query with initial timestamp.

daigorowhite opened the pull request:
In case of large table and not needed all record, or user just need recent record and realtime data from now, initial query is required with specific timestamp. but currently, initial query uses `Timestamp(0L)`.

In this PR , `kafka-connect-jdbc`'s `source` have new config which is `timestamp.initial`.
if you specify `timestamp.initial`, it works like this. 

- `0` : `1970-01-01 00:00:00.000` (Same as existed)
- `-1` : Current time will be used 
- `1528858322371` : `2018-06-13 11:52:02.371` ( UnixTime milliseconds )

and default is 0.

In our case, we need this function, could you have a look, please?

daigorowhite commented:
[clabot:check]

daigorowhite commented:
hi @rhauch 
I think this use case is needed for some users,
since this issue has same topic for `incrementing` mode. ( our PR is currently only for `timestamp`)
https://github.com/confluentinc/kafka-connect-jdbc/issues/432

If you have time , please have a look this PR.

thanks

rhauch commented on src/test/java/io/confluent/connect/jdbc/source/JdbcSourceTaskUpdateTest.java:
```diff
@@ -472,6 +551,14 @@ private void startTask(String timestampColumn, String incrementingColumn, String
+    startTask(timestampColumn, incrementingColumn, query, delay, timestampInitial, null);
+  }
+
+  private void startTask(String timestampColumn, String incrementingColumn, String query, Long delay, Long timestampInitial, Long incrementInitial) {
```
Is `incrementInitial` an artifact of trying to do something similar for the incremental offset? It's currently unused, so we should remove it or add support for incremental offsets, too.

rhauch commented on docs/source_config_options.rst:
```diff
@@ -130,6 +130,13 @@ Mode
+  * Type: long
+  * Default: 0
+  * Importance: low
+
```
This file was removed from the codebase.

rhauch commented on src/main/java/io/confluent/connect/jdbc/source/JdbcSourceConnectorConfig.java:
```diff
@@ -165,6 +165,14 @@
   private static final String TIMESTAMP_COLUMN_NAME_DISPLAY = "Timestamp Column Name";
 
+  public static final String TIMESTAMP_INITIAL_CONFIG = "timestamp.initial";
+  public static final long TIMESTAMP_INITIAL_DEFAULT = 0;
```
Should the default be `null`? That seems to be better than `0` if we want this to be an optional property.

rhauch commented on src/main/java/io/confluent/connect/jdbc/source/JdbcSourceConnectorConfig.java:
```diff
@@ -165,6 +165,14 @@
+  public static final long TIMESTAMP_INITIAL_CURRENT = -1;
+  public static final String TIMESTAMP_INITIAL_DOC =
+      "Long value to specify when the query offset starts."
+      + "offset starts from current time if the value is -1.";
```
How about:
```
"The epoch timestamp used for initial queries that use timestamp criteria. "
      + "Use -1 to use the current time. If not specified, all data will be retrieved."
```

rhauch commented on src/main/java/io/confluent/connect/jdbc/source/JdbcSourceTask.java:
```diff
@@ -163,7 +166,10 @@ public void start(Map<String, String> properties) {
+
       Map<String, Object> offset = offsets == null ? null : offsets.get(partition);
+      offset = offset == null ? fetchInitialOffset(timestampInitial) : offset;
+
```
Actually, we should use `dialect.currentTimeOnDB(connection, calendar)` to get the current time in the database. Also, be sure to see the recent changes from #505 that allow the user to specify the time zone, and that will affect the calendar that's used.

daigorowhite commented:
@rhauch thanks for having review!
I have handled your comment and modify the code.
please check again ? and sorry, I have no idea about jenkins checkstyle...
could you tell me how to fix it ?

daigorowhite commented:
Let me try to remind again @rhauch 
Thanks

rgogulakrishnan commented:
Hi, any update on this feature..? When this will be rolled out..

daigorowhite commented:
hi , I would like to merge this since I got approval already. no update now, I am just waiting :)

remis87 commented:
Any updates on this feature yet? Waiting for it passionately

aliasbadwolf commented:
@daigorowhite Although I like the idea to handle it through configuration (and I am super late to comment) but the same can be done by setting the initial timestamp in the offset topic for connect framework. Before starting the connector, just add an entry to the offset topic with whatever timestamp you want to run the initial query from (and this can handle both timestamp and incremental scenarios). Although its a hack, until your changes get merged.

HunderlineK commented:
@rhauch @wicknicks Could you please review the PR again and merge it if it looks okay? Thanks

sohaibfarooqi commented:
Any updates on this? Waiting for this feature for quite some time!

gharris1727 commented on src/main/java/io/confluent/connect/jdbc/source/JdbcSourceTask.java:
```diff
@@ -256,6 +260,36 @@ public void start(Map<String, String> properties) {
+        // start at the specified timestamp
+        if (timestampInitial == JdbcSourceConnectorConfig.TIMESTAMP_INITIAL_CURRENT) {
+          // use the current time
+          try (Connection con = cachedConnectionProvider.getConnection()) {
```
Can you pull the `getConnection` outside of the try-with-resources?
It's a cached connection object, and closing it here invalidates that object elsewhere and forces the task to reconnect before continuing. See how `validateNonNullable` does it for an example.

daigorowhite replied:
got it, I moved this one. Could you check my change if I understand your comment ?

daigorowhite commented on src/main/java/io/confluent/connect/jdbc/source/JdbcSourceTask.java:
```diff
@@ -38,16 +46,9 @@
-import io.confluent.connect.jdbc.util.ColumnId;
-import io.confluent.connect.jdbc.util.TableId;
-import io.confluent.connect.jdbc.util.Version;
-
```
I reverted.

daigorowhite commented on src/main/java/io/confluent/connect/jdbc/source/JdbcSourceTask.java:
```diff
@@ -149,6 +150,8 @@ public void start(Map<String, String> properties) {
     boolean validateNonNulls
         = config.getBoolean(JdbcSourceTaskConfig.VALIDATE_NON_NULL_CONFIG);
+    Long timestampInitial
+        = config.getLong(JdbcSourceConnectorConfig.TIMESTAMP_INITIAL_CONFIG);
```
I removed this one .

daigorowhite commented:
@gharris1727 
Hi, Thank you for your kind review! I have fixed with your comment.
and, I switched destination `5.5.x` .
please kindly review again?
